src/Logging.jl

```diff
@@ -86,25 +127,47 @@ show(io::IO, logger::Logger) = print(io, "Logger(", join([logger.name,
 
-const _root = Logger("root", WARNING, STDERR)
-Logger(name::AbstractString;args...) = configure(Logger(name, WARNING, [STDERR], _root); args...)
+const _root = Logger("root", WARNING, LogIO(STDERR))
+Logger(name::AbstractString; kwargs...) = configure(Logger(name, WARNING, [LogIO(STDERR)], _root); kwargs...)
 Logger() = Logger("logger")
 
 write_log(syslog::SysLog, color::Symbol, msg::AbstractString) = send(syslog.socket, syslog.ip, syslog.port, length(msg) > syslog.maxlength ? msg[1:syslog.maxlength] : msg)
-write_log{T<:IO}(output::T, color::Symbol, msg::AbstractString) = (print(output, msg); flush(output))
+write_log(output::IO, color::Symbol, msg::AbstractString) = (print(output, msg); flush(output))
 write_log(output::Base.TTY, color::Symbol, msg::AbstractString) = Base.print_with_color(color, output, msg)
 
-function log(syslog::SysLog, level::LogLevel, color::Symbol, logger_name::AbstractString, msg...)
-    # syslog needs a timestamp in the form: YYYY-MM-DDTHH:MM:SS-TZ:TZ
-    t = time()
-    timestamp = string(strftime("%Y-%m-%dT%H:%M:%S",t), strftime("%z",t)[1:end-2], ":", strftime("%z",t)[end-1:end])
-    logstring = string("<", ((@compat UInt16(syslog.facility)) << 3) + (@compat UInt16(level)), ">1 ", timestamp, " ", syslog.machine, " ", syslog.user, " - - - ", level, ":", logger_name,":", msg...)
-    write_log(syslog, color, logstring)
+function formatlog(format::AbstractString, logging_fields::Dict)
+    function substr(s)
+        v = get(logging_fields, s[3:end-1], "<none>")
+        return isa(v, Function) ? string(v()) : string(v)
+    end
+    return replace(format, r"%\([^)]+\)", substr)
+end
+
+function formatlog(logoutput::LogOutput, level::LogLevel, loggername::AbstractString, timestamp::ZonedDateTime, msg...)
+    logoutput.logging_fields["msg"] = string(msg...)
+    logoutput.logging_fields["timestamp"] = Dates.format(ZonedDateTime(timestamp, logoutput.timezone), logoutput.timestamp_format)
+    logoutput.logging_fields["level"] = level
+    logoutput.logging_fields["loggername"] = loggername
+    return formatlog(logoutput.format, logoutput.logging_fields)
+end
+
+function log(output::LogIO, level::LogLevel, color::Symbol, loggername::AbstractString, msg...)
+    timestamp = TimeZones.now(UTC)
+    write_log(output.io, color, formatlog(output, level, loggername, timestamp, msg...))
+end
+
+function syslog_facility_level(facility, level)
+    return string("<", (UInt16(facility) << 3) + UInt16(level), ">1")
 end
 
-function log{T<:IO}(output::T, level::LogLevel, color::Symbol, logger_name::AbstractString, msg...)
-    logstring = string(strftime("%d-%b %H:%M:%S",time()),":",level, ":",logger_name,":", msg...,"\n")
-    write_log(output, color, logstring)
+function log(syslog::SysLog, level::LogLevel, color::Symbol, loggername::AbstractString, msg...)
+    timestamp = TimeZones.now(UTC)
+    syslog_timestamp = Dates.format(timestamp, syslog_timestamp_format)
```

Comment:
Here you could annotate the global like `syslog_timestamp_format::AbstractString` to avoid the performance hit of globals. I won't pretend to know for sure that it would affect you here, though.